core/src/main/java/org/apache/iceberg/SnapshotManager.java

```diff
@@ -0,0 +1,146 @@
+  }
+
+  @Override
+  protected ManageSnapshots self() {
+    return this;
+  }
+
+  @Override
+  protected String operation() {
+    Snapshot manageSnapshot = base.snapshot(targetSnapshotId);
+    return manageSnapshot.operation();
+  }
+
+  @Override
+  public ManageSnapshots cherrypick(long snapshotId) {
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null,
+        "Cannot cherry pick unknown snapshot id: %s", snapshotId);
+
+    operation = SnapshotManagerOperation.CHERRYPICK;
+    this.targetSnapshotId = snapshotId;
+    return this;
+  }
+
+  @Override
+  public ManageSnapshots rollback(long snapshotId) {
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null,
+        "Cannot roll back to unknown snapshot id: %s", snapshotId);
+
+    operation = SnapshotManagerOperation.ROLLBACK;
+    this.targetSnapshotId = snapshotId;
+    return this;
+  }
+
+  @Override
+  public ManageSnapshots rollbackToTime(long timestampMillis) {
+    long snapshotId = 0;
+    long snapshotTimestamp = 0;
+    operation = SnapshotManagerOperation.ROLLBACK;
+    // find the latest snapshot by timestamp older than timestampMillis
+    for (Snapshot snapshot : base.snapshots()) {
+      if (snapshot.timestampMillis() < timestampMillis &&
+          snapshot.timestampMillis() > snapshotTimestamp) {
+        snapshotId = snapshot.snapshotId();
+        snapshotTimestamp = snapshot.timestampMillis();
+      }
+    }
+
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null,
+        "Cannot roll back, no valid snapshot older than: %s", timestampMillis);
+
+    this.targetSnapshotId = snapshotId;
+
+    return this;
+  }
+
+  @Override
+  public Snapshot apply() {
+    // common checks
+    ValidationException.check(targetSnapshotId != null,
+        "Cannot run operations on an unknown version: call rollback, rollbackAtTime or cherrypick");
+    ValidationException.check(operation != null,
+        "Need to define operation on snapshot: call rollback, rollbackAtTime or cherrypick");
+
+    switch (operation) {
+      case CHERRYPICK:
+        Snapshot cherryPickSnapshot = base.snapshot(this.targetSnapshotId);
+        String wapId = stagedWapId(cherryPickSnapshot);
+        boolean isWapWorkflow =  wapId != null && !"".equals(wapId);
+        if (isWapWorkflow) {
+          ValidationException.check(!base.isWapIdPublished(wapId),
+              "Duplicate request to cherry pick wap id that was published already: %s", wapId);
+        }
+        // only append operations are currently supported
+        if (!cherryPickSnapshot.operation().equals(DataOperations.APPEND)) {
+          throw new UnsupportedOperationException("Can cherry pick only append operations");
+        }
+        // this is to ensure we add files only once for each targetSnapshotId, to protect from
+        // duplicate additions if commit retries the same cherrypick operation on failure.
+        if (!snapshotsAlreadyCherrypicked.contains(targetSnapshotId)) {
```

Comment:
on testing this feature we found that when there is high concurrent writing, `ops.commit()` can fail and get retried by the `SnapshotProducer.commit()` .. this requires Cherrypick operation to be idempotent. Since adding files changes state of the base table metadata we need to ensure this is done exactly once.